src/libos/crates/async-rt/src/time/instant.rs

```diff
@@ -0,0 +1,125 @@
+use crate::prelude::*;
+use core::ops::{Add, AddAssign, Sub, SubAssign};
+use vdso_time::{clock_gettime, ClockId};
+
+lazy_static! {
+    static ref LAST_NOW: Mutex<Instant> = Mutex::new(Instant(DURATION_ZERO));
+}
+
+pub const DURATION_ZERO: Duration = Duration::from_nanos(0);
+
+/// A measurement of a monotonically nondecreasing clock. Opaque and useful only with Duration.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Instant(pub(crate) Duration);
+
+impl Instant {
+    /// Returns an instant corresponding to “now”.
+    ///
+    /// Rust std says that the clock of linux + x86 environment is real-monotonic.
+    /// When sgx feature enabled, we are in linux + x86 env, the clock should be real-monotonic.
+    /// However, the clock in sgx env is untrusted, we can not guarantee that clock is real-monotonic.
+    /// To ensure the instant is monotonically nondecreasing, We keep a global "latest now" instance
+    /// which is returned instead of what the OS says if the OS goes backwards.
+    pub fn now() -> Self {
+        let os_now = Instant(clock_gettime(ClockId::CLOCK_MONOTONIC).unwrap());
+
+        let mut last_now = LAST_NOW.lock();
+        let now = core::cmp::max(*last_now, os_now);
+        *last_now = now;
```

Comment:
This implementation is good. A straightforward (and buggy) implementation would check `os_now` against `LAST_NOW`, which could result in false-positive failures in some race conditions involving concurrent updates to `LAST_NOW`.